[sparse][semi-structured][inductor] meta registrations for _cslt_sparse_mm + additional stride checking in test.

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114685

_cslt_sparse_mm + additional stride checking in test.

Summary:

This PR adds in meta registrations for _cslt_sparse_mm.

Based on the work @drisspg did
in #114370.

Additionally, it updates the tests by checking that the strides of the
spare result and the result returned by sparse+compile are the same, to
avoid errors like those found in

https://github.com/pytorch/pytorch/pull/114477.

Test Plan:
```
python test/test_sparse_semi_structred -k compile_cusparselt
python test/test_sparse_semi_structred -k compile_cutlass
```

Reviewers:

Subscribers:

Tasks:

Tags:

jcaip committed:
[sparse][semi-structured][inductor] meta registrations for
_cslt_sparse_mm + additional stride checking in test.

Summary:

This PR adds in meta registrations for _cslt_sparse_mm.

Based on the work @drisspg did
in #114370.

Additionally, it updates the tests by checking that the strides of the
spare result and the result returned by sparse+compile are the same, to
avoid errors like those found in

https://github.com/pytorch/pytorch/pull/114477.

Test Plan:
```
python test/test_sparse_semi_structred -k compile_cusparselt
python test/test_sparse_semi_structred -k compile_cutlass
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

danthe3rd commented on test/test_sparse_semi_structured.py:
```diff
@@ -146,28 +146,46 @@ def forward(self, x):
-            for backend in SEMI_STRUCTURED_SUPPORTED_BACKENDS:
-                for dense_input_shape in [(1, 128), (64, 128), (128, 128), (64, 128, 128)]:
-                    _test_mlp_contiguous_relu_compile(backend, dense_input_shape)
+    @unittest.skipIf(IS_WINDOWS, "torch.compile not support on windows")
+    @unittest.skipIf("cusparselt" not in SEMI_STRUCTURED_SUPPORTED_BACKENDS, "cusparselt not supported on this machine")
+    def test_mlp_contiguous_relu_compile_cusparselt(self):
+        """
+        test for cuSPASRELt meta registrations (_cslt_sparse_mm) + torch.compile
+        """
+        for dense_input_shape in [(1, 128), (64, 128), (128, 128), (64, 128, 128)]:
+            SparseSemiStructuredTensorCompileTest._test_mlp_contiguous_relu_compile("cusparselt", dense_input_shape)
+
+    @unittest.skipIf(IS_WINDOWS, "torch.compile not support on windows")
+    def test_mlp_contiguous_relu_compile_cutlass(self):
+        """
+        test for CUTLASS meta registrations (_sparse_semi_structured_linear) + torch.compile
+        """
+        for dense_input_shape in [(1, 128), (64, 128), (128, 128), (64, 128, 128)]:
+            SparseSemiStructuredTensorCompileTest._test_mlp_contiguous_relu_compile("cutlass", dense_input_shape)
```
nit: You should be able to use `@parametrize` to "sweep" over the input shape / backend.

danthe3rd commented on test/test_sparse_semi_structured.py:
```diff
@@ -146,28 +146,46 @@ def forward(self, x):
-        def _test_mlp_contiguous_relu_compile(backend, dense_input_shape):
-            SparseSemiStructuredTensor._FORCE_CUTLASS = backend == "cutlass"
-            input = torch.rand(dense_input_shape, device="cuda").half()
+        SparseSemiStructuredTensor._FORCE_CUTLASS = backend == "cutlass"
```
Shouldn't this be done in a context manager sort of thing?
This will modify this value for all subsequent tests

jcaip replied:
Yes, each test currently sets this though so it won't cause an error. I plan to do some refactoring work splitting out the cuSPASRELt and CUTLASS subclasses, will address the flags in the tests then.

drisspg commented on torch/_meta_registrations.py:
```diff
@@ -394,6 +394,29 @@ def meta_unsqueeze_(self, dim):
+    out_dtype: Optional[torch.dtype] = None,
+    transpose_result: bool = False,
+):
+    assert len(dense_B.shape) == 2, "cslt_sparse_mm only supports 2d inputs"
```
Maybe replicate the two other checks found in the C++ func

drisspg reviewed (commented):
I think it looks good modulo Daniels comments. If you wanted to get much more rigours checking for you, you can add a opinfo in common_method_invocations for the aten op and this will trigger many tests, one being the fake_tensor checking which ensures the meta registration is correct on your sample inputs

alexsamardzic reviewed (approved):
LGTM.

jcaip committed:
Update on "[sparse][semi-structured][inductor] meta registrations for _cslt_sparse_mm + additional stride checking in test."


_cslt_sparse_mm + additional stride checking in test.

Summary:

This PR adds in meta registrations for _cslt_sparse_mm.

Based on the work drisspg did
in #114370.

Additionally, it updates the tests by checking that the strides of the
spare result and the result returned by sparse+compile are the same, to
avoid errors like those found in

https://github.com/pytorch/pytorch/pull/114477.

Test Plan:
```
python test/test_sparse_semi_structred -k compile_cusparselt
python test/test_sparse_semi_structred -k compile_cutlass
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][semi-structured][inductor] meta registrations for _cslt_sparse_mm + additional stride checking in test."


_cslt_sparse_mm + additional stride checking in test.

Summary:

This PR adds in meta registrations for _cslt_sparse_mm.

Based on the work drisspg did
in #114370.

Additionally, it updates the tests by checking that the strides of the
spare result and the result returned by sparse+compile are the same, to
avoid errors like those found in

https://github.com/pytorch/pytorch/pull/114477.

Test Plan:
```
python test/test_sparse_semi_structred -k compile_cusparselt
python test/test_sparse_semi_structred -k compile_cutlass
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
@pytorchbot merge

jcaip commented:
@pytorchbot merge
